src/shared_modules/indexer_connector/src/indexerConnectorSyncImpl.hpp

```diff
@@ -664,6 +664,52 @@ class IndexerConnectorSyncImpl final
         return resultJson;
     }
 
+    void executeSearchQueryWithPagination(const std::string& index,
+                                         const nlohmann::json& query,
+                                         std::function<void(const nlohmann::json&)> onResponse = nullptr)
+    {
+        nlohmann::json currentQuery = query;
+        std::string searchAfter;
+        while (true)
+        {
+            nlohmann::json searchResult = executeSearchQuery(index, currentQuery);
+
+            // Always call the callback, even for empty pages, to notify the caller of each page's result (including empty pages).
+            if (onResponse)
+            {
+                onResponse(searchResult);
+            }
+
+            // If first page and empty, still break after callback
+            if (!searchResult.contains("hits") || !searchResult["hits"].contains("hits") ||
```

Comment:
The triple `contains()` check followed by `empty()` is noisy and still relies on `operator[]` internally. Use `find()` + pointer access (`get_ptr`) to reliably test existence and type without side effects.

Example

```cpp
auto itHits = searchResult.find("hits");
if (itHits == searchResult.end()) { break; }

auto itInner = itHits->find("hits");
if (itInner == itHits->end() || !itInner->is_array() || itInner->empty()) { break; }
```

This hasn't accidental mutation, it fast and predictable.